### PR TITLE
Added Tests for Search Pipeline and Notifications Plugin

### LINKED ISSUE #474: [FEATURE] Add support for search pipeline APIs
### Is your feature request related to a problem?

The [search pipeline feature](https://opensearch.org/docs/latest/search-plugins/search-pipelines/index/) is GA in version 2.9.0 and we would like to get them supported.

There is an issue with border decription. https://github.com/opensearch-project/opensearch-py/issues/435

We are looking for more descriptive interface and well defined parameters for the high level pipelines APIs. The interface could be modeled the pipeline as an object with a description, id, version, and lists of processors. The processors would be objects with types, tags, descriptions and "config". The config would just be modeled as a map from string to object.

Without modification, search pipeline CRUD must be built using REST API. Ref: https://opensearch-project.github.io/opensearch-py/api-ref/transport.html

The client side interface of the search pipeline could be consistent with that of the ingest pipeline. Ref https://opensearch-project.github.io/opensearch-py/api-ref/clients/ingest_client.html

This is prerequisite to add benchmarks in opensearch-benchmark repo. https://github.com/opensearch-project/OpenSearch/issues/7782#issuecomment-1679371986

### PR DESCRIPTION
**Pull request by AbitraryYu**
### Description
_Describe what this change achieves._

This change tries to solve #474 

### Issues Resolved
_List any issues this PR will resolve, e.g. Closes [...]._ 

If this is resolved, this closes #474 

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

**Comment by orangewise** [comment]
Question, would this also work with opensearch serverless (aws)?

**Review by dblock — CHANGES_REQUESTED** [review]
Was this code actually generated? The "right" way to do this is to add search pipelines to https://github.com/opensearch-project/opensearch-api-specification and then to generate the client.

**Comment by AbitraryYu** [comment]
I see there's a [PR in opensearch-api-specification](https://github.com/opensearch-project/opensearch-api-specification/pull/172) . If that PR is merged, then does that mean the search pipeline api for opensearch-py can be generated? Thanks.

**Comment by saimedhi** [comment]
> I see there's a [PR in opensearch-api-specification](https://github.com/opensearch-project/opensearch-api-specification/pull/172) . If that PR is merged, then does that mean the search pipeline api for opensearch-py can be generated? Thanks.

- @AbitraryYu, once the spec is merged, we can generate the search pipeline API. After generation, we'll review it for any necessary fixes to the generator code.

- Your contribution to getting the spec merged would be greatly appreciated, and it will help us move forward with API generation. Thank you.

**Comment by lambda-science** [comment]
Any follow up on this ? .. We need search pipeline  API

**Comment by saimedhi** [comment]
> Any follow up on this ? .. We need search pipeline API


@lambda-science, @ArbitraryYu, please feel free to contribute to the spec to get it merged.

https://github.com/opensearch-project/opensearch-api-specification/pull/172

**Comment by AbitraryYu** [comment]
When I try to generate the code using `nox -rs generate` it fails at `/utils/generate_api.py` line 595. I have printed some of the debug logs (by inserting some print statements on my own) and saw the following:

```
{'name': 'config_id', 'in': 'query', 'description': 'A channel ID.', 'schema': {'type': 'string', 'description': 'A channel ID.'}}
{'name': 'config_id_list', 'in': 'query', 'description': 'A comma-separated list of channel IDs.', 'schema': {'type': 'string', 'description': 'A comma-separated list of channel IDs.'}}
{'name': 'last_updated_time_ms', 'in': 'query', 'schema': {'type': 'integer', 'format': 'int64'}}
Traceback (most recent call last):
  File "<repo_dir>/utils/generate_api.py", line 853, in <module>
    dump_modules(read_modules())
                 ^^^^^^^^^^^^^^
  File "<repo_dir>/utils/generate_api.py", line 595, in read_modules
    type=m["schema"]["type"], description=m["description"]
                                          ~^^^^^^^^^^^^^^^
KeyError: 'description'
nox > Command python utils/generate_api.py failed with exit code 1
nox > Session generate failed.
```

I take a look at the `OpenSearch.openapi.json` at raw.github, and find out it is the GET parameters of `/_plugins/_notifications/configs` that lack the description (on line 13377, or see the picture)
2024-03-26-110252_1920x1051_scrot

I saw some other parameters also lack description, I am not sure if there are more parameters that lack description.

My question is: Are there any methods that I can skip the problematic part and just generate the search pipeline API part?If not, is it a new issue that need to be raised in the opensearch-api-specification repository?

**Comment by saimedhi** [comment]
> When I try to generate the code using `nox -rs generate` it fails at `/utils/generate_api.py` line 595. I have printed some of the debug logs (by inserting some print statements on my own) and saw the following:
> 
> ```
> {'name': 'config_id', 'in': 'query', 'description': 'A channel ID.', 'schema': {'type': 'string', 'description': 'A channel ID.'}}
> {'name': 'config_id_list', 'in': 'query', 'description': 'A comma-separated list of channel IDs.', 'schema': {'type': 'string', 'description': 'A comma-separated list of channel IDs.'}}
> {'name': 'last_updated_time_ms', 'in': 'query', 'schema': {'type': 'integer', 'format': 'int64'}}
> Traceback (most recent call last):
>   File "<repo_dir>/utils/generate_api.py", line 853, in <module>
>     dump_modules(read_modules())
>                  ^^^^^^^^^^^^^^
>   File "<repo_dir>/utils/generate_api.py", line 595, in read_modules
>     type=m["schema"]["type"], description=m["description"]
>                                           ~^^^^^^^^^^^^^^^
> KeyError: 'description'
> nox > Command python utils/generate_api.py failed with exit code 1
> nox > Session generate failed.
> ```
> 
> I take a look at the `OpenSearch.openapi.json` at raw.github, and find out it is the GET parameters of `/_plugins/_notifications/configs` that lack the description (on line 13377, or see the picture) 2024-03-26-110252_1920x1051_scrot
> 
> I saw some other parameters also lack description, I am not sure if there are more parameters that lack description.
> 
> My question is: Are there any methods that I can skip the problematic part and just generate the search pipeline API part?If not, is it a new issue that need to be raised in the opensearch-api-specification repository?



> When I try to generate the code using `nox -rs generate` it fails at `/utils/generate_api.py` line 595. I have printed some of the debug logs (by inserting some print statements on my own) and saw the following:
> 
> ```
> {'name': 'config_id', 'in': 'query', 'description': 'A channel ID.', 'schema': {'type': 'string', 'description': 'A channel ID.'}}
> {'name': 'config_id_list', 'in': 'query', 'description': 'A comma-separated list of channel IDs.', 'schema': {'type': 'string', 'description': 'A comma-separated list of channel IDs.'}}
> {'name': 'last_updated_time_ms', 'in': 'query', 'schema': {'type': 'integer', 'format': 'int64'}}
> Traceback (most recent call last):
>   File "<repo_dir>/utils/generate_api.py", line 853, in <module>
>     dump_modules(read_modules())
>                  ^^^^^^^^^^^^^^
>   File "<repo_dir>/utils/generate_api.py", line 595, in read_modules
>     type=m["schema"]["type"], description=m["description"]
>                                           ~^^^^^^^^^^^^^^^
> KeyError: 'description'
> nox > Command python utils/generate_api.py failed with exit code 1
> nox > Session generate failed.
> ```
> 
> I take a look at the `OpenSearch.openapi.json` at raw.github, and find out it is the GET parameters of `/_plugins/_notifications/configs` that lack the description (on line 13377, or see the picture) 2024-03-26-110252_1920x1051_scrot
> 
> I saw some other parameters also lack description, I am not sure if there are more parameters that lack description.
> 
> My question is: Are there any methods that I can skip the problematic part and just generate the search pipeline API part?If not, is it a new issue that need to be raised in the opensearch-api-specification repository?

Hey @AbitraryYu, to avoid the generator error, we could tweak the code to first check if the parameter description exists before using it. Yet, I believe it's even better to include the parameter description that is missing in the spec.

@dblock, what's your take on this?

**Comment by saimedhi** [comment]
@AbitraryYu, please try fixing  https://github.com/opensearch-project/opensearch-py/issues/710.

So, that all APIs from spec will be generated by [update_api workflow](https://github.com/opensearch-project/opensearch-py/blob/main/.github/workflows/update_api.yml) and PR will be raised if changes detected.  Consider raising PR for spec repo with missing descriptions. Thank you.

**Comment by AbitraryYu** [comment]
The code is generated via nox, and I modified the generator to gracefully ignore messages if the API specification is empty.

By the way, the raw link of the opensearch "https://raw.githubusercontent.com/opensearch-project/opensearch-api-specification/main/OpenSearch.openapi.json" seems giving me 404 (line 550-554), so I use my local version downloaded from last week.

**Comment by AbitraryYu** [comment]
A side note, the latest version of opensearch (docker) require admin password, and I cannot get it run tests. I am not sure if this repository supports password parsing so I am using an older version: 2.11.1 (docker).

File: utils/generate_api.py
Comment (dblock) [inline utils/generate_api.py:600]: Make it the same as below for `default`, create `a` (really could use a better variable name) first, then `update` it with `description` if any,

**Review by dblock — CHANGES_REQUESTED** [review]
Let me try to help with the password issue.

In 2.12 you need a default password, https://opensearch.org/blog/replacing-default-admin-credentials/, this repo was updated to act accordingly AFAIK.

We're moving things around in opensearch-api-specification, so you might hit issues, https://github.com/opensearch-project/opensearch-api-specification/issues/189

Iterate this change to green?

File: utils/generate_api.py
Comment (saimedhi) [inline utils/generate_api.py:601]: @AbitraryYu, could you handle [params description](https://github.com/opensearch-project/opensearch-py/blob/main/utils/generate_api.py#L593) similarly to how parts description is handled [here](https://github.com/opensearch-project/opensearch-py/blob/main/utils/generate_api.py#L630)?

```
                a = dict(type=m["schema"]["type"])  # pylint: disable=invalid-name

                if "description" in m:
                    a.update({"description": m["description"]})

```

**Comment by saimedhi** [comment]
@AbitraryYu, Thank you for implementing the requested changes! Could you also include tests for these generated APIs?

**Comment by saimedhi** [comment]
> Hi @saimedhi , I got confused where I should implement the tests. I saw `test_opensearch/test_helpers/test_search.py` and `test_opensearch/test_async/test_helpers/test_search.py`, and they are using `from opensearchpy.helpers import search`. I am wondering do I need to create a new helper file in opensearchpy.helpers plus creating both new test files in `test_opensearch/test_helpers/` and `test_opensearch//test_async/test_helpers/`. Thanks.

@AbitraryYu, 

- For search_pipeline unit tests, they can be added [here](https://github.com/opensearch-project/opensearch-py/tree/main/test_opensearchpy/test_client). Please create a file for search_pipeline test cases in that folder.
- For search_pipeline integration tests, removing the skip test [here](https://github.com/opensearch-project/opensearch-py/blob/main/test_opensearchpy/test_server/test_rest_api_spec.py#L75) is sufficient.
- Notification plugin tests can be added [here](https://github.com/opensearch-project/opensearch-py/tree/main/test_opensearchpy/test_server/test_plugins), and async tests [here](https://github.com/opensearch-project/opensearch-py/tree/main/test_opensearchpy/test_async/test_server/test_plugins).

**Comment by AbitraryYu** [comment]
Hi @saimedhi , I just updated with the tests by following this [notification api link](https://opensearch.org/docs/latest/observing-your-data/notifications/api/#update-channel-configuration) and this [search pipeline link](https://opensearch.org/docs/latest/search-plugins/search-pipelines/creating-search-pipeline/)

**Comment by AbitraryYu** [comment]
Sorry for all the dirty commits, I accidentally committed some sensitive data to this branch and I attempt to remove it both locally and remotely. I already force push the git history so the my local repo feature branch cannot access it. However, this PR still have the stray commit that contain the sensitive data.

I just submitted a Github support request for to clear the cached views but probably need this PR's admin permission to actually take effect.

File: opensearchpy/_async/client/__init__.py
Comment (saimedhi) [inline opensearchpy/_async/client/__init__.py:227]: @ArbitraryYu, thanks for your contribution. Just a few adjustments needed. 

- No modification needed for this file. In a recent PR, `self.search_pipeline = SearchPipelineClient(self)` was added.

File: opensearchpy/client/__init__.py
Comment (saimedhi) [inline opensearchpy/client/__init__.py:227]: No modification needed for this file

File: test_opensearchpy/test_server/test_rest_api_spec.py
Comment (saimedhi) [inline test_opensearchpy/test_server/test_rest_api_spec.py:75]: Please remove the change `OpenSearch-main/rest-api-spec/src/main/resources/rest-api-spec/test/search_pipeline/10_basic` to avoid blocking this PR, as two tests are currently failing. 

It will be handled in #726 . Integration tests with OpenSearch 2.7 and 2.8 are failing due to a server bug, and I'll raise an issue in the appropriate repo.

**Review by saimedhi — CHANGES_REQUESTED** [review]

**Comment by saimedhi** [comment]
@ArbitraryYu, please resolve the failing DCO check.

**Comment by AbitraryYu** [comment]
For the DCO, I try to follow the DCO guide, `git rebase HEAD~10 --signoff`, going through the rebase process, and once I push it, the PR got 49+ files changed. I was wondering if I can add those signoff(s) by modifying a specific commit?

**Comment by dblock** [comment]
> For the DCO, I try to follow the DCO guide, `git rebase HEAD~10 --signoff`, going through the rebase process, and once I push it, the PR got 49+ files changed. I was wondering if I can add those signoff(s) by modifying a specific commit?

It's probably easier to re-create the changes in the branch, or do https://code.dblock.org/2015/08/31/getting-out-of-your-first-git-mess.html. 

```
git checkout feature/search_pipeline_api
git checkout -b feature/search_pipeline_api-backup
git checkout feature/search_pipeline_api
git reset HEAD~1 # repeat this until you've rewound your branch enough
git add .
git commit -s -m ...
git push origin feature/search_pipeline_api-backup -f
```

**Comment by AbitraryYu** [comment]
Fixed the DCO, and the tests seems ok.

**Review by saimedhi — APPROVED** [review]

File: test_opensearchpy/test_server/test_plugins/test_notification.py
Comment (saimedhi) [inline test_opensearchpy/test_server/test_plugins/test_notification.py:17]: - @AbitraryYu, please remove async from all the test cases in this file.
- After removing async, you might notice that some test cases fail. Please fix them.
- For example, test_delete_channel_configuration may fail. Remember, each test case is isolated, so create the configuration before deleting it within the same test case. See [this reference](https://github.com/opensearch-project/opensearch-py/blob/main/test_opensearchpy/test_server/test_plugins/test_index_management.py#L108). Thank you.

**Review by saimedhi — CHANGES_REQUESTED** [review]

**Comment by saimedhi** [comment]
@AbitraryYu, please try to fix failing tests.

**Comment by AbitraryYu** [comment]
I followed the example on test_index_management.py of [this reference](https://github.com/opensearch-project/opensearch-py/blob/main/test_opensearchpy/test_server/test_plugins/test_index_management.py#L97C1-L98C69), and I got errors like `error: Unsupported target for indexed assignment ("Collection[str]")` on L96-97,L100-101 and `error: Value of type "Collection[str]" is not indexable` on L102. How can I fix that?

Meanwhile, the `list_all_config()` (which calls `/_plugins/_notifications/configs`) returns an empty list of config list, after I created the config; while `test_get_config()` (which calls `/_plugins/_notifications/configs/<config_id>`) return something on the config list. I am not sure if  `list_all_config()` `/_plugins/_notifications/configs` is behaving properly.

**Comment by saimedhi** [comment]
> I followed the example on test_index_management.py of [this reference](https://github.com/opensearch-project/opensearch-py/blob/main/test_opensearchpy/test_server/test_plugins/test_index_management.py#L97C1-L98C69), and I got errors like:
> 
> * `error: Unsupported target for indexed assignment ("Collection[str]")` on [here](https://github.com/AbitraryYu/opensearch-py/blob/feature/search_pipeline_api/test_opensearchpy/test_server/test_plugins/test_notification.py#L95C1-L98C55)
> * `error: Value of type "Collection[str]" is not indexable` on [here.](https://github.com/AbitraryYu/opensearch-py/blob/feature/search_pipeline_api/test_opensearchpy/test_server/test_plugins/test_notification.py#L99C9-L99C89)
> 
> How can I fix that?

@ArbitraryYu, follow these steps to resolve the errors:

- Format the code using `nox -rs format`.
- Add type hints for CONTENT [here](https://github.com/ArbitraryYu/opensearch-py/blob/feature/search_pipeline_api/test_opensearchpy/test_server/test_plugins/test_notification.py#L24):

```
CONTENT: Dict[str, Any] = {
    "config_id": "sample-id",
    # ... other keys ...
}
```

**Review by saimedhi — APPROVED** [review]

**Comment by saimedhi** [comment]
@dblock, I'm unable to merge the PR. Could you please review, approve, and merge it? Thank you!

**Review by dblock — APPROVED** [review]

**Pull request merged**
